yamux/src/session.rs

```diff
@@ -679,3 +679,167 @@ impl<T> Drop for Session<T> {
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::Session;
+    use crate::{
+        config::Config,
+        frame::{Flag, Flags, Frame, FrameCodec, GoAwayCode, Type},
+    };
+    use futures::{
+        channel::mpsc::{channel, Receiver, Sender},
+        stream::FusedStream,
+        SinkExt, Stream, StreamExt,
+    };
+    use std::{
+        io,
+        pin::Pin,
+        task::{Context, Poll},
+    };
+    use tokio::{
+        io::AsyncReadExt,
+        prelude::{AsyncRead, AsyncWrite},
+    };
+    use tokio_util::codec::Framed;
+
+    struct MockSock {
+        sender: Sender<Vec<u8>>,
+        receiver: Receiver<Vec<u8>>,
+        read_buffer: Vec<u8>,
+    }
+
+    impl MockSock {
+        fn new() -> (Self, Self) {
+            let (tx, rx) = channel(25);
+            let (tx_1, rx_1) = channel(25);
+
+            (
+                MockSock {
+                    sender: tx,
+                    receiver: rx_1,
+                    read_buffer: Default::default(),
+                },
+                MockSock {
+                    sender: tx_1,
+                    receiver: rx,
+                    read_buffer: Default::default(),
+                },
+            )
+        }
+    }
+
+    impl AsyncRead for MockSock {
+        fn poll_read(
+            mut self: Pin<&mut Self>,
+            cx: &mut Context,
+            buf: &mut [u8],
+        ) -> Poll<io::Result<usize>> {
+            loop {
+                if self.receiver.is_terminated() {
+                    break;
+                }
+                match Pin::new(&mut self.receiver).poll_next(cx) {
+                    Poll::Ready(Some(data)) => self.read_buffer.extend(data),
+                    Poll::Ready(None) => {
+                        return Poll::Ready(Err(io::ErrorKind::BrokenPipe.into()));
+                    }
+                    Poll::Pending => break,
+                }
+            }
+
+            let n = ::std::cmp::min(buf.len(), self.read_buffer.len());
+
+            if n == 0 {
+                Poll::Pending
+            } else {
+                buf[..n].copy_from_slice(&self.read_buffer[..n]);
+                self.read_buffer.drain(..n);
+                Poll::Ready(Ok(n))
+            }
+        }
+    }
+
+    impl AsyncWrite for MockSock {
+        fn poll_write(
+            mut self: Pin<&mut Self>,
+            cx: &mut Context,
+            buf: &[u8],
+        ) -> Poll<io::Result<usize>> {
+            match self.sender.poll_ready(cx) {
+                Poll::Ready(Ok(())) => match self.sender.try_send(buf.to_vec()) {
```

Comment:
To simulate the real situation, I think the MockSock can have an option to control the max bytes can be written. However, this may not be needed in tests here.